lib/greek_stemmer.rb

```diff
@@ -39,12 +39,7 @@ def stem(word)
     stem = word.dup
     return stem if PROTECTED_WORDS.include?(stem) || !greek?(word)
 
-    step_1_regexp = /(.*)(#{STEP_1_EXCEPTIONS.keys.join("|")})$/u
-
-    # step 1
-    stem.scan(step_1_regexp) do |st, suffix|
-      stem = st + STEP_1_EXCEPTIONS[suffix]
-    end
+    return STEP_1_EXCEPTIONS[stem] if STEP_1_EXCEPTIONS[stem]
```

Comment:
My thoughts:

1. Ok this makes sense. My guess is that it was originally intended as a catch-all case for words with common suffixes. However it probably degenerated into full-word exceptions as well.
2. I also agree that we should probably add another step. One handles full words (and we put only the hardest cases there) and the other one that handles suffix-only exceptions.
3. We should first check the size of the changes on our corpus.
4. Before merging, we should estimate the difficulty vs. gain of upgrading the Java plugin.

Good catch though!